pkg/controller/azure/node/predicate.go

```diff
@@ -15,50 +15,68 @@
 package node
 
 import (
+	"time"
+
 	"github.com/go-logr/logr"
 	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/cache"
 	"sigs.k8s.io/controller-runtime/pkg/event"
 	"sigs.k8s.io/controller-runtime/pkg/predicate"
 )
 
-// NewReadyUnreachableChangedPredicate creates a new predicate that filters only relevant node events,
-// such as node creation and deletion, updating the deletion timestamp of a node, changes to the "Ready" condition,
-// and changes to the "Unreachable" taint.
-func NewReadyUnreachableChangedPredicate(logger logr.Logger) predicate.Predicate {
+const (
+	nodeCacheTTL = 10 * time.Hour
+)
+
+// NewNodePredicate creates a new predicate that filters only relevant node events,
+// such as creating or deleting a node, updating the deletion timestamp of a node,
+// or updating the ready condition or unreachable taint of a node.
+func NewNodePredicate(logger logr.Logger) predicate.Predicate {
+	nodeCache := cache.NewExpiring()
 	return predicate.Funcs{
 		CreateFunc: func(e event.CreateEvent) bool {
 			if e.Object == nil {
 				logger.Error(nil, "CreateEvent has no object", "event", e)
 				return false
 			}
-			if _, ok := e.Object.(*corev1.Node); !ok {
+			node, ok := e.Object.(*corev1.Node)
+			if !ok {
 				return false
 			}
-			logger := logger.WithValues("name", e.Object.GetName())
+			logger := logger.WithValues("name", node.Name)
 			logger.Info("Creating a node")
+			nodeCache.Set(node.Name, node, nodeCacheTTL)
```

Comment:
As said, I don't think having a cache in the predicates is strictly necessary. 
With respect to caching node objects, it might even be suboptimal, as the can become quite large and tend to be updated frequently. That means, your cache might grow fast and the predicate invocations are contending with each other for the cache's lock. 
I don't want to say it's a bad idea, but you might need to be careful regarding scalability of this approach.

Reply:
Thanks, this is really a good point! Nevertheless, I already tried other approaches to deal with missed events but I didn't like the result. Trying to determine if an event should result in a reconciliation based on looking at virtual machines in the client cache is possible, but is much more complex and results in unneeded reconciliations that can't be avoided. By contrast, using an expiring cache is simple and elegant, so I would like to stay with it if at all possible.

> can become quite large and tend to be updated frequently. That means, your cache might grow fast and the predicate invocations are contending with each other for the cache's lock.

* The cache can contain at most all node objects of shoots scheduled on the seed (it's expiring so if a node is removed, it will be eventually removed from the cache as well). With a max of 250 shoots and an average of 20 nodes per shoot (that's probably an overestimation), it could grow to about 5000 objects. A YAML representation of a node is about 10K (binary should be less), so the max cache size should be about 50M, nothing dramatic in my opinion. 
* A node receives hearbeat updates every 10s, so this means max 500 node updates per second, or 1 every 2 milliseconds. I agree this is a bit too much and can result in lock contention if the cache is updated on every event. 
 
In terms of lock contention, the expensive part is `Set` (`Get` uses `RLock / RUnlock` which should be cheap). Therefore, it could be avoided by not updating the node cache on every event, e.g. by skipping these very frequent hearbeat events or all events that don't result in changes we are interested in. This should reduce the lock contention to very tolerable levels even with thousands of objects in the cache. Let me work on this a bit more.

Reply:
I introduced a simple but IMO effective solution to the lock contention issue by not updating the cache on events caused by changes we are not interested in, PTAL.

Reply:
Yes, that's a nice idea :)

Reply:
I realized the above calculation is flawed, since the remedy controller is deployed (unlike extensions) to the control plane of each shoot, and therefore cares about a single shoot, not for all shoots in a seed. Therefore, we should be interested in the max number of nodes a single shoot can have, instead of the average number of nodes per shoot.

I checked our largest landscape. The largest number of nodes I was able to find for a single shoot was 228, their size in yaml is 3022204, so ~3M, not ~50M as I originally calculated. For ~200 nodes, the frequency of 10s events would be 20 nodes per second, so 1 node every 50 ms, which is already much more tolerable.

Nevertheless, I still think it makes sense to be careful about both the cache size and the frequency of cache updates, so I am still proposing to keep the change I already did. In addition, I realized we don't need to keep entire node objects in the cache, but only "projections" containing the information we are interested in. Let me modify the PR a bit in this direction.

Reply:
I added 3 more commits that:
* Change the node and service caches to use projections rather than full objects
* Refactor the node and service predicates for testability
* Add unit tests to the node and service predicates

@timebertt I would really appreciate another look if you have time. I am perhaps spending too much time on this, but this is because I am using this PR as a case study for the proper way to handle missed events in such predicates, potentially not only for the remedy controller but also for other projects.

Reply:
/unhold

Reply:
> I am perhaps spending too much time on this, but this is because I am using this PR as a case study for the proper way to handle missed events in such predicates, potentially not only for the remedy controller but also for other projects.

No worries, I think this time is definitely well-spent.
I appreciate that you consider all of these details here, it will be valuable for other projects in the future as well 👍